### DISCUSSION TITLE
Large-scale refactoring & supporting Objective-C and Swift

### DISCUSSION THREAD

**Discussion by amorde** [MobileNativeFoundation/discussions#24] [General]
Hello!

Square has run into several hurdles while performing large-scale refactors across our monorepo. We have a large amount of Objective-C and Swift and Xcode's built-in refactoring is not reliable enough nor does it support more complex refactoring.

We are starting to build some of our own tooling but wondering what other folks are doing to solve this.

Some examples of desired refactors:

* Add a new parameter to an initializer (think low level framework used by hundreds of modules)
* Change the type of an existing parameter
* Change the name of an existing parameter
* Swap out one type for another, handling the various ways classes types are referenced in each language (ex. `SomeClass *` vs. `id<SomeProtocol>` vs `SomeProtocol` in Swift)
* Perform a replacement within a specific context (ex. when inside a function named `someFunc`, replace `X` with `Y`)

With all of the above able to handle APIs defined in each language and consumed by each language.

We have looked at things like [codemod](https://github.com/facebook/codemod) but that still requires manually building regular expressions and it would be difficult to build generic workflows for adding new parameters.

**Comment by alephao** [comment]
Hello! 

Three years ago I was also migrating a codebase from Obj-C to Swift. As an interim step, I wrote some python scripts to make the Objective-C API nicer to use in Swift code.

Those scripts were just regex substitutions with extras to handle some edge cases, for example, one of the scripts would add [NS_SWIFT_NAME](https://developer.apple.com/documentation/swift/objective-c_and_c_code_customization/renaming_objective-c_apis_for_swift) to some Objective-C properties.

Unfortunately, I couldn't find the scripts specifically to the Obj-C to Swift migration, but here is one example of a script that helped migrating some XCTest to Nimbe matchers. It follows the same idea of the Obj-C to Swift ones we had. Hopefully it can help you somehow: https://gist.github.com/alephao/55f0e98ecd29796d45cb35b6d6569112

**Comment by BalestraPatrick** [comment]
Xcode's built-in refactoring is also not enough for the few big refactors I had to do in the past in our project. I found Xcode's search and replace regex support to be pretty good though if you didn't know this feature existed (sorry for the shameless plug): https://patrickbalestra.com/blog/2020/02/09/xcode-find-navigator.html

**Reply by amorde** [reply]
Xcode's regex support is pretty good, agreed! Our biggest issue is that we normally work on isolated subprojects that contain only portions of the entire monorepo, and opening a workspace containing everything leaves Xcode fairly unresponsive. We've started writing scripts that do regex replacements outside of Xcode.

**Reply by BalestraPatrick** [reply]
We also have some kind of focused projects, and as you can see in my blog post, there's a way to actually specify a custom search scope in Xcode. In this way, you can run search/replace inside of Xcode for code not in the IDE.

**Comment by rudro** [comment]
Been meaning to try this but haven't had a chance, seems quite capable : https://comby.dev

**Reply by amorde** [reply]
This looks really interesting! Thank you for sharing

**Comment by aaroncrespo** [comment]
Its kinda wonky and Xcode centric (not very automated) but some types of migrations can be handled by writing availability attributes on the old renaming to the new: `@availability(*, deprecated, message: "", renamed: "apiNext(_:)") func oldAPI(...) {` 

You select all the warnings and tell Xcode to accept the fix-its. This doesn't really work with arity changes, but at least you know where all your call sites are..

Relatedly maybe a tool that that could interpret fix-its generated by a build log and apply them the same way Xcode does is something  that could bridge some gaps. We can write some types of fix-it generating code, we know the compiler is always adding more.
